### DISCUSSION TITLE
Material property transfer in THM

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#23349] [Q&A General] [answered]
The VariableValueTransferMaterial can create an AD material property for a variable transferred from the boundary of a 2D mesh onto a 1D mesh. And now I have a material property which I want to transfer from the boundary of a 2D mesh onto a 1D mesh. How can I achieve it? Should I rewrite VariableValueTransferMaterial.C and make it possible to transfer the material property from the boundary of a 2D mesh onto a 1D mesh? @joshuahansel 

Ethan

**Comment by Ethan-xj** [comment]
By the way, in VariableValueTransferMaterial
```
ChannelVariableTransferMaterial::computeProperties()
{
  std::vector<ADReal> T_w_nodal_values;
  for (const auto i : _current_elem->node_index_range())
  {
    const Node & nd = _current_elem->node_ref(i);

    // Assumes the variable you are coupling to is from the nonlinear system for now.
    const Node * const nearest = _nearest_node.nearestNode(nd.id());
    mooseAssert(nearest, "I do not have the nearest node for you");
    const auto dof_number = nearest->dof_number(_nl_sys.number(), _paired_variable, 0);
    T_w_nodal_values.push_back((*_serialized_solution)(dof_number));
    Moose::derivInsert(T_w_nodal_values.back().derivatives(), dof_number, 1.);
  }
```
What is `(*_serialized_solution)(dof_number)`? I don't know much about it.

**Reply by GiudGiud** [reply]
@joshuahansel

**Comment by joshuahansel** [comment, marked as answer]
@lindsayad `VariableValueTransferMaterial` is hot right now. What do you think about being able to transfer material properties rather than variables?

Ethan, at the very least, if you have no other option, in principle, you can transfer all of the material property's variable dependencies onto the 1D side and then recompute it there, but I'm not sure in your case how complex that would be.

**Reply by lindsayad** [reply]
How would you evaluate the material property from the node?

**Reply by joshuahansel** [reply]
Right, it couldn't be like it is now with the nearest node. I meant to ask if there was some kind of nearest QP transfer we could do 😄 Or just any other idea to solve this problem. Obviously there's also the maybe terrible solution I suggested above. Or AD aux variables 🤤

**Reply by lindsayad** [reply]
Evaluating the variable dependencies seems like a pretty good idea to me. This kind of thing would be pretty easy with a functor material property, but I doubt @Ethan-xj is familiar with those (we'd also have to add a nodal-evaluation overload as well, so it wouldn't work right out of the box ... it would work out of the box with a nearest-qp evaluation)

**Reply by Ethan-xj** [reply]
I found a new problem. The variable transfered by `VariableValueTransferMaterial` is nonlinear. Can `VariableValueTransferMaterial` transfer aux variables?

**Reply by Ethan-xj** [reply]
As for the material property I want to transfer, I think I can use the `VariableValueTransferAux` to transfer variable dependencies and recompute the material property on the 1D side.

**Reply by joshuahansel** [reply]
@lindsayad can check my work here: I expect that `VariableValueTransferMaterial` could be modified to work with aux variables; however, you would lose the AD in that chain (anytime you introduce aux variables into your calculation). You'd probably be better off recomputing the aux quantities on the 1D side if you can manage that.

**Reply by Ethan-xj** [reply]
I‘m working on the lateral flow between 2 1D flow channels and I want to transfer the density velocity and pressure of one channel to another. But in FlowChannelSinglePhase pressure is a aux variable. So in current situation I can’t transfer pressure in AD form，right？

**Reply by joshuahansel** [reply]
Right, you can't transfer pressure directly. You'll need to transfer the solution variables `rhoA`, `rhouA`, and `rhoEA` and then recompute pressure. See `ADFluidProperties3EqnMaterial` for how to compute pressure from these.

**Reply by joshuahansel** [reply]
(In your case, if you're using `VariableValueTransferMaterial`, those solution variables will come into your new material as material properties instead of coupled variables.)

**Reply by lindsayad** [reply]
We could definitely modify that material to work with aux variables if desired

**Reply by Ethan-xj** [reply]
I think I've transfered `rhoA`, `rhouA` and `rhoEA` and recompute the property by the help of `ADFluidProperties3EqnMaterial`. But I realized that I didn't solve the original problem. The material property I want to transfer is the gradient of the velocity(I firstly create a new material class which compute the gradient of the velocity by the variable `rhoA` `rhouA` and their gradient). Now I can transfer the `rhoA`, `rhouA` and `rhoEA` in two ways: material property and aux variable. In material property, I can't use `adCoupledGradient` to express the gradient of `rhoA`, `rhouA`. And in aux variable, as you say, it would lose the AD form. 
I'm working on the convection term in the lateral flow equation. That's the reason why I need to express the gradient of the velocity(both 2 channels'). THM uses the DGKernel to deal with the convection term, but I can't fully understand the `calcFlux()` in `ADNumericalFlux3EqnHLLC` Do you have any good suggestions?

**Reply by joshuahansel** [reply]
> I think I've transfered rhoA, rhouA and rhoEA and recompute the property by the help of `ADFluidProperties3EqnMaterial`.

I don't understand how this is possible: with `VariableValueTransferMaterial`, you have `rhoA`, `rhouA`, and `rhoEA` as material properties, not variables, which is what `ADFluidProperties3EqnMaterial` uses... I was suggesting that you make a new material that just copies how `ADFluidProperties3EqnMaterial` computes the pressure material property.

> The material property I want to transfer is the gradient of the velocity

Yes, this does make things trickier. I think it's doable though. Here's what I'm thinking:
```
u = rhouA / rhoA
grad(u) = grad(rhouA / rhoA)
grad(u) = grad(rhouA) / rhoA + rhouA * grad(1/rhoA)
grad(u) = grad(rhouA) / rhoA - rhouA/(rhoA)^2 * grad(rhoA)
```
So basically, you need to have `rhoA`, `rhouA`, `grad(rhoA)`, and `grad(rhouA)` on hand to compute your velocity gradient. The first two you can get with `VariableValueTransferMaterial`. The second two will need you to make a sister class like `VariableGradientTransferMaterial` to compute the gradient instead of the value (This makes me think we should do some refactoring to reduce code duplication, but that can be done later...). Lastly you'll have another material to do your velocity gradient calcuation (`ADParsedMaterial` should work).

Let me know if that makes sense.

**Reply by Ethan-xj** [reply]
> I don't understand how this is possible: with VariableValueTransferMaterial, you have rhoA, rhouA, and rhoEA as material properties, not variables, which is what ADFluidProperties3EqnMaterial uses... I was suggesting that you make a new material that just copies how ADFluidProperties3EqnMaterial computes the pressure material property.

Yes, my statement is not accurate. Not the real help, I just copy the way how `ADFluidProperties3EqnMaterial` computes the material property.
> So basically, you need to have rhoA, rhouA, grad(rhoA), and grad(rhouA) on hand to compute your velocity gradient. The first two you can get with VariableValueTransferMaterial. The second two will need you to make a sister class like VariableGradientTransferMaterial to compute the gradient instead of the value (This makes me think we should do some refactoring to reduce code duplication, but that can be done later...). Lastly you'll have another material to do your velocity gradient calcuation (ADParsedMaterial should work).

Thank you for your advice and I'll have a try.

**Reply by Ethan-xj** [reply]
There are some difficulties:
1. I'm not very familiar with base code in moose. In other words, I've no idea about how to compute the gradient instead of the value. I found the `gradSln()` function but I don't know if it's useful and if I can use it in 'VariableGradientTransferMaterial' straightly.
2. If I complish the VariableGradientTransferMaterial, how should I test it?

**Reply by joshuahansel** [reply]
I've made you a branch with the new material: https://github.com/joshuahansel/moose/tree/var-gradient-transfer-material. You can cherry pick the last two commits.

Note that I haven't done any testing or documentation on this (I've only checked that it compiles), so these wouldn't be able to be merged into MOOSE immediately. Try it out and see if it works, and if it does, we can go from there.

**Reply by Ethan-xj** [reply]
That really helps me a lot :D
By the way, I add a new variable so I should use `addBoundaryCondition` to set the bc. But 1D mesh's boundary is node. How can I get the inlet or outlet node id of the 1D THM mesh?

**Reply by joshuahansel** [reply]
> By the way, I add a new variable so I should use addBoundaryCondition to set the bc. But 1D mesh's boundary is node. How can I get the inlet or outlet node id of the 1D THM mesh?

I do not understand this. What `addBoundaryCondition` are you talking about? To clarify, even though it's 1D, you still have the option of using an integrated (weak) boundary condition instead of a nodal (strong, Dirichlet) boundary condition. But to answer your question, you should have the nodesets `mychannel:in` and `mychannel:out` available to you, where `mychannel` is the name of your channel. This is what you would supply to the `boundary` parameter of your BC.
